### ISSUE TITLE
API for Assets to get the Engine Version?

### ISSUE THREAD

**Issue by BenjamenMeyer** [vegastrike/Vega-Strike-Engine-Source#349]
Q: Do we have an API for the Game Assets to be able to get the VS Engine version?
If not, we should add one.

Uses:
- Games can report the version of the engine they're using in addition to the version of the game itself; reducing the need to update the game assets whenever the game engine changes.
- Games can verify that the game engine is the required version.

**Comment by BenjamenMeyer** [comment]
This seems to be a duplicate of #152

**Comment by BenjamenMeyer** [comment]
#152 is about the internal variables and storage of version information.
This is about opening that up to be exposed to the Game Assets so they can query the engine version and be able to determine if they're compatible with the engine or not.

**Comment by BenjamenMeyer** [comment]
per gitter.im conversation - let's get this into 0.8.x.
0.9.x is listed to ensure we get it into master too.

**Comment by BenjamenMeyer** [comment]
Python API should look something like:
```
version = VS.version()
```
If one outputs `version` under a debugger/Python IDLE:
```
>>> print(version)
[0, 9, 0]
```
If one needs a text version, then they can use `str.join`:
```
>>> strVersion = ".".join(version)
>>> print(strVersion)
"0.9.0"
```
Not sure how to accomplish this with our current Python integration.

**Comment by ermo** [comment]
I assume that we would need to check this as the very first thing when trying to start an asset set?

E.g.


```python
def check_engine_version():
    asset_title = "Vega Strike: Upon the Coldest Sea"
    asset_api_version = (0, 8, 0)
    engine_api_version = VS.version()
    print("%s (API version %s) running on the Vega Strike engine (API version %s)", asset_title, ".".join(asset_version), ".".join(engine_version))
    if asset_version != engine_version:
        debug.fatal("ERROR: The current asset API version (%s) is not compatible with the running VS engine API version (%s).", ".".join(asset_version), ".".join(engine_version))
        exit(1)
```

**Comment by stephengtuggy** [comment]
@ermo Yes, something like that. Only I expect we would just compare the first and second numbers of the tuple. Make sure that the first number equaled the expected value, and that the second was at least the expected value, if not greater. Along the lines of [SemVer](https://semver.org/) .

**Comment by BenjamenMeyer** [comment]
@ermo  yes something like that, though you can also use other comparisons as tuples make for great means:
```python
>>> x = (0,9,0)
>>> y = (0,10,0)
>>> y > x
True
```

So:
```python
min_version_required = (0, 8, 0)
game_asset_name = "your game name"
def check_engine_version():
    engine_api_version = VS.version()
    if engine_api_version < min_version_required:
        debug.fatal(f"ERROR: {game_asset_name} requires Vega Strike Engine {".".join(min_version_required)} or newer; but found Vega Strike Engine {".".join(engine_api_version)} instead.")
        sys.exit(1)
```

There may be a case for checking maximum versions too:
```python
engine_api_version > max_version_allowed
```
Though that would require more up-keep but would be just as simple.

**Comment by BenjamenMeyer** [comment]
https://www.boost.org/doc/libs/develop/libs/tuple/doc/html/tuple_users_guide.html
Looks useful

**Comment by stephengtuggy** [comment]
We still need to implement this, don't we? In fact, I think that the Engine Version API might be the biggest single thing holding us back from releasing 0.8.x at this point.

**Comment by BenjamenMeyer** [comment]
yes, sorry - I've been busy with other stuff IRL but it will be a key point for 0.8.x

**Comment by BenjamenMeyer** [comment]
starting work on this; also centralizing the version information into the CMake file so it's all auto-generated from there and we only have one place in the code to update going forward.

Starting on master, will then port to 0.8.x

**Comment by stephengtuggy** [comment]
@BenjamenMeyer Cool.

Note that, in #557 , I got the git SHA to be populated in the CMake file. We may want to merge that before you proceed with your changes.

**Comment by BenjamenMeyer** [comment]
After thinking about this a bit more...

1. We do want to have the Engine Version (e.g 0.8.0, 0.9.0, etc) exposed to the Game Assets.
2. We also need to have an Assets API Version.

The first provides the information about the engine and is useful for showing it to the user, etc.

The second is something I'm taking from Google Android and how they version their API for applications; it's simple and concise and may move independently of the system itself. This can be useful to us as we may have several releases that use the same Asset API Version. It also means that if need be we can rev it for a minor release if we need to fix a bug.

I've already got the data exposed in the existing version locations so it's all centralized:
```
me@myself:~/Devel/Vega-Strike-Engine-Source/build$ ./vegastrike-engine --version
[2021-09-26 15:31:52.746279] [0x00007f09bbcd8d40] [trace]   Registering codec ogg
[2021-09-26 15:31:52.746318] [0x00007f09bbcd8d40] [trace]   .
Legacy Mode: FALSE
 In path /home/me/Devel/Vega-Strike-Engine-Source/build
Vega Strike  
See http://www.gnu.org/copyleft/gpl.html for license details.

ARG #1 = --version
Vega Strike Engine Version 0.9.0.124984680 ( Assets API Version: 1)

me@myself:~/Devel/Vega-Strike-Engine-Source/build$
```

@stephengtuggy yes, this does depend on having the Git Hash available, so we'll need to have that ported to 0.8.x to get this onto 0.8.x as well.

Now to figure out how to add the Python APIs...

**Comment by BenjamenMeyer** [comment]
Update: Have some code that will pull in the version data, and added an interface to hook it up to the Python side. The Engine compiles, but having issues with the Game Assets actually using it. Need to learn the Engine Python API side a little more.

**Comment by BenjamenMeyer** [comment]
filed PR #578 with my current work; need some help getting the new object to be part of the Python `VS` object.

**Comment by stephengtuggy** [comment]
This is done now on both 0.8.x and master, right, @BenjamenMeyer ?

**Comment by BenjamenMeyer** [comment]
Yes this is fully done as far as code goes. We should start recording changes for the various versions as we go. So may be we should start an engine API version readme doc before closing this entirely.

**Comment by BenjamenMeyer** [comment]
Updated the description with the 4 required tasks.

**Comment by BenjamenMeyer** [comment]
Filed #617 against 0.8.x to introduce the asset API change log. Copied code from VS:UtCS for detecting the Asset API version for example code in the document.

**Comment by BenjamenMeyer** [comment]
Wiki Page for documentation https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Asset-API
